### PR TITLE
vdk-trino: collect lineage for select/insert and rename table only

### PR DESCRIPTION
**Pull request by philip-alexiev**
Why:
To make lineage collecting more production ready,
some improvements are needed.

What:
In order to reduce the load on the query engine,
   only plans for insert/select queries are calculated.
For rename table queries, the plan doesn't give information.
   The query is parsed and table names extracted.
Counting the number of rows in the output table before and after
   is removed to reduce the burden on the query engine.

How has this been tested:
Tweaked the test_vdk_trino_lineage.py test
  to be more comprehensive and cover all scenarios.

What type of change are you making?
Bug fix (non-breaking change which fixes an issue)
  or a cosmetic change/minor improvement

Signed-off-by: Philip Alexiev (palexiev@vmware.com)

### CODE REVIEW

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py [hunk 1]
```diff
@@ -109,61 +107,46 @@ def execute_query_with_retries(self, query):
...
-                with closing_noexcept_on_close(self.__cursor()) as cur:
+                with closing_noexcept_on_close(self._cursor()) as cur:
                     cur.execute(f"EXPLAIN (TYPE IO, FORMAT JSON) {query}")
                     result = cur.fetchall()
                     if result:
                         import json
 
                         data = json.loads(result[0][0])
-                        data["@type"] = "taurus_query_io"
+                        data["@type"] = "insert_select"
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:138]: Why not specify the concrete type returned by sqlparse - "select" "insert" , and above "alter" ?

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py [hunk 2]
```diff
@@ -109,61 +107,46 @@ def execute_query_with_retries(self, query):
...
-        if self._lineage_logger:
+
+        from vdk.plugin.trino import lineage_utils
+        import sqlparse
+
+        statement = sqlparse.parse(query)[0]
+
+        if statement.get_type() == "ALTER":
+            rename_table_names = lineage_utils.parse_rename_table_names(query)
+            if rename_table_names:
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:119]: let's add logs (debug) especially in the non covered branches (what if rename table_names is None) ?

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 3]
```diff
@@ -27,26 +31,133 @@
...
+                and lineage_data["@type"] == "insert_select"
+                and lineage_data["status"] == "OK"
+                and lineage_data["query"] == insert_query
+                and re.fullmatch(r"[0-9]{10}\.[0-9]+", lineage_data["@id"])
+                and lineage_data["outputTable"]["schemaTable"]["table"] == TEST_TABLE_1
+            )
+
+    mock_lineage_logger.send.assert_called_with(InsertLineageDataMatch())
+
+    #   Test select query
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:79]: why not split it into multiple test methods ? Having one long test method is generally not a best idea since 
A) it's harder to read (having long methods is not best practice, it's good to have some rule keep methods below N lines)
B) it's less parallelizable
C) it takes more time to troubleshoot (e.g if multiple test fails you discover them one by one.

**Review by tozka — CHANGES_REQUESTED** [review]

**Comment by tozka** [comment]
I am not sure if you noticed - the CI tests failed (ci/gitlab/gitlab.com) - https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/2176218096

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py (hunk 1 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:138]: There's one more type  "insert into <table> select from <table>" which is a combination. So basically insert and select queries can be considered an insert_select statement with no input or no output. Just wanted to make things simpler. Do you think it makes sense to have strictly support for 4 operations:
- insert
- select
- insert_select
- rename_table (alter is too generic)

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py (hunk 2 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:119]: OK

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 3 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:79]: I thought about that and it's cleaner. Although, each test spawns a new Trino container, as I understand, so it will make the execution a lot heavier (although more parallelizable). Do you think I should split?

Comment (tozka) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:79]: The trino fixture is "session" scope so it should be spawned only once per test session . So it should not be a problem.

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py (hunk 1 above)
Comment (tozka) [reply projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:138]: We can have insert select (`INSERT INTO orders SELECT * FROM new_orders;`
or insert values (`INSERT INTO cities VALUES (1, 'San Francisco');`) 

You want to differentiate them ?

Comment (tozka) [reply projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:138]: And yes we can support (for now) those operations only

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 3 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:79]: Great. I will split them.

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 4]
```diff
@@ -26,27 +28,225 @@
...
+    result: Result = runner.invoke(["trino-query", "--query", insert_query])
+    cli_assert_equal(0, result)
+
+    # the lineage data is different on every run of the test,
+    # so we need this class to generalize the dict which is to be matched
+    class InsertLineageDataMatch:
+        def __eq__(self, lineage_data):
+            import json
+
+            print("INSERT query lineage data:")
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:59]: remove the prints?

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 5]
```diff
@@ -26,27 +28,225 @@
...
+
+    table_name = "test_table_" + uuid.uuid4().hex
+
+    mock_lineage_logger = mock.MagicMock(LineageLogger)
+
+    class TestConfigPlugin:
+        @hookimpl
+        def vdk_initialize(self, context):
+            context.state.set(LINEAGE_LOGGER_KEY, mock_lineage_logger)
+
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:93]: why not extract the class at file level and re-use it (you'd pass the lineage logger as a constructor argument)

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 6]
```diff
@@ -26,27 +28,225 @@
...
+    cli_assert_equal(0, result)
+
+    result: Result = runner.invoke(
+        [
+            "trino-query",
+            "--query",
+            f"create table {table_name_to} (test_column varchar)",
+        ]
+    )
+    cli_assert_equal(0, result)
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:167]: nit: you can create helper method def execute_query (query)

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 7]
```diff
@@ -26,27 +28,225 @@
...
+    result: Result = runner.invoke(
+        [
+            "trino-query",
+            "--query",
+            f"create table {table_name_from} (test_column varchar)",
+        ]
+    )
+    cli_assert_equal(0, result)
+
+    rename_query = f"alter table {table_name_from} rename to {table_name_to}"
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:229]: We should include some "negative" tests. For example what if teh query is "alter view" or "alter table add column" or "describe"

File: projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py [hunk 8]
```diff
@@ -109,61 +107,48 @@ def execute_query_with_retries(self, query):
...
-
     def _get_lineage_data(self, query, query_id):
-        if self._lineage_logger:
+
+        from vdk.plugin.trino import lineage_utils
+        import sqlparse
+
+        statement = sqlparse.parse(query)[0]
+
+        if statement.get_type() == "ALTER":
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/src/vdk/plugin/trino/trino_connection.py:117]: Below I asked about "alter table add column" test because I wonder what would happen here?

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 9]
```diff
@@ -26,27 +28,225 @@
...
+            "trino-query",
+            "--query",
+            f"create table {table_name_to} (test_column varchar)",
+        ]
+    )
+    cli_assert_equal(0, result)
 
-    class LineageDataMatch:  # the lineage data is different on every run of the test,
-        # so we need this class to generalize the dict which is to be matched
+    insert_select_query = f"insert into {table_name_to} select * from {table_name_from}"
```
Comment (tozka) [inline projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:169]: Maybe add tests that include fully qualified names (catalog.schema.table) and (schema.table) 

?

**Review by tozka — APPROVED** [review]
Looks good to me. You can add a few more tests about some corner cases.

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 5 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:93]: Already done that in my next change :)

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 4 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:59]: Was wondering about that. May be useful when debugging, but then again, you can run the tests locally and add the print. I will remove them.

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 7 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:229]: I will. Also, I will add a test for "alter table IF EXISTS <tbl> rename to <tbl>"

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 9 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:169]: I am handling those in my next change. Brace yourself, it's a big change :D

**Comment by philip-alexiev** [comment]
@tozka Thank you for the review and valuable comments.

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py (hunk 7 above)
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py:229]: Unfortunately, the memory connector is limited so it doesn't seem possible to make "rename column" or "add column" queries.  The test will be limited. DESCRIBE is supported.

**Review by tozka — APPROVED** [review]

File: projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py [hunk 10]
```diff
@@ -26,27 +97,263 @@
...
+
+@pytest.mark.usefixtures("trino_service")
+@mock.patch.dict(
+    os.environ,
+    {
+        VDK_DB_DEFAULT_TYPE: "TRINO",
+        VDK_TRINO_PORT: "8080",
+        VDK_TRINO_USE_SSL: "False",
+    },
+)
```
[the remark this answers is not in the document]
Comment (philip-alexiev) [reply projects/vdk-plugins/vdk-trino/tests/test_vdk_trino_lineage.py]: Great. Thanks.
